### ISSUE TITLE
[TypeScript] Add support for typings in EventDefinition

### ISSUE THREAD

**Issue by Itrulia** [rangle/redux-beacon#301]
This is a...
----
 - [ ] :beetle: Bug Report
 - [x] :rocket: Feature Request
 - [ ] :scroll: Documentation Request

:rocket: :scroll: What's missing from Redux Beacon that you'd like to add?
----
Currently to set a typing to the `action` parameter of `EventDefinition` you will need to do it similar to this:

```
const pageView = trackPageView(a => {
    const action: LocationChangeAction = a as any;

    return {
        page: `${action.payload.pathname}${action.payload.search}`,
        location: action.payload.pathname,
    };
});
```

this is because 

```
{
    [key: string]: any;
}
```

is not compatible with anything that looks different. I would suggest using generics and change it to

```
export type EventDefinition<E = any, A = { [key: string]: any }, S = any> = (
  action: A,
  prevState: S,
  nextState: S
) => E;
```

This will keep it backwards compatible. But it also requires to change the package typings.

E.g. of GoogleAnalytics package

```
export declare const trackPageView: <A = { [key: string]: any }, S = any>(eventDef: EventDefinition<{
    page: string;
    title?: string;
    location?: string;
}, A, S>, tracker?: string | string[]) => EventDefinition<any, A, S>;
```

Can you help out?
----
 - [x] :star2: I am a legend and can get started on a pull request right away given the go-ahead.
 - [ ] :star: I am a superstar and would like to help out given some guidance.
 - [ ] :disappointed: I won't be able to help out on this one.

**Comment by ttmarek** [comment]
I've been meaning to tighten up the types here for a while. I think it's definitely a priority at this point. Let's start with `types.ts` in the main `redux-beacon` package`. Extend the generics for `EventDefinition` like you have there...that could be one pr, then once that's in we can update the event definition helpers in each target. Naturally we'd start with whichever one you're using (e.g. GA).

This is a great idea, and I'll love to get this done.

**Comment by Itrulia** [comment]
Will take a look on the weekend!

**Comment by ttmarek** [comment]
Everything discussed here was beautifully implemented and is now available in `@redux-beacon/google-analytics@1.2.0` and `redux-beacon@2.0.5`. Thanks!
